File: Doc-Scanner-X/Doc-Scanner-X/UploadedFileViewController.m
//
//  UploadedFileViewController.m
//  Doc-Scanner-X
//
//  Created by dynamsoft on 2018/1/9.
//  Copyright © 2018年 com.dynamsoft. All rights reserved.
//

#import "UploadedFileViewController.h"

@interface UploadedFileViewController ()

@end

@implementation UploadedFileViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _webView = [[WKWebView alloc] initWithFrame:CGRectZero configuration:[WKWebViewConfiguration new]];
    _webView.allowsBackForwardNavigationGestures = YES;
    _webView.navigationDelegate = self;
    _webView.frame = CGRectMake(0, 64, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height-64);
    
    _navigationBar = [[CustomNavigationBar alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, 64)];
    self.navigationItem.title = @"Uploaded Files";
    [_navigationBar pushNavigationItem:self.navigationItem animated:YES];
    MainViewController* parentVC = (MainViewController*)self.parentViewController;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"icon_back"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] style:UIBarButtonItemStylePlain target:parentVC action:@selector(backToHome)];
    
    _refreshControl = [UIRefreshControl new];
    _refreshControl.attributedTitle = [[NSAttributedString alloc] initWithString:@"Pull to refresh"];
    [_refreshControl addTarget:self action:@selector(refreshWeb) forControlEvents:UIControlEventValueChanged];
    _webView.scrollView.bounces = YES;
    [_webView.scrollView addSubview:_refreshControl];
    
    self.view.frame = CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height);
    [self.view addSubview:_navigationBar];
    [self.view addSubview:_webView];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [_webView addObserver:self forKeyPath:@"canGoBack" options:NSKeyValueObservingOptionNew context:nil];
    NSURL* myURL = [[NSURL alloc] initWithString:[@"https://demo.dynamsoft.com/DCS_Mobile/filesList.html?userId=" stringByAppendingString:[KeyChainManager readUUID]]];
    NSURLRequest* myRequest = [[NSURLRequest alloc] initWithURL:myURL];
    [_webView loadRequest:myRequest];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [_webView removeObserver:self forKeyPath:@"canGoBack"];
}

- (void)refreshWeb {
    [_refreshControl endRefreshing];
    NSURL* myURL = [[NSURL alloc] initWithString:[@"https://demo.dynamsoft.com/DCS_Mobile/filesList.html?userId=" stringByAppendingString:[KeyChainManager readUUID]]];
    NSURLRequest* myRequest = [[NSURLRequest alloc] initWithURL:myURL];
    [_webView loadRequest:myRequest];
}

- (void)back {
    [_webView goBack];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if (_webView.canGoBack) {
        self.navigationItem.leftBarButtonItem.target = self;
        self.navigationItem.leftBarButtonItem.action = @selector(back);
    } else {
        MainViewController* parentVC = (MainViewController*)self.parentViewController;
        self.navigationItem.leftBarButtonItem.target = parentVC;
        self.navigationItem.leftBarButtonItem.action = @selector(backToHome);
    }
}

- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(WKNavigation *)navigation {
    _activityIndicator.backgroundColor = [UIColor clearColor];
    _activityIndicator.center = CGPointMake([UIScreen mainScreen].bounds.size.width/2, [UIScreen mainScreen].bounds.size.height/2);
    _activityIndicator.hidesWhenStopped = YES;
    [self.view addSubview:_activityIndicator];
    [_activityIndicator startAnimating];
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    [_activityIndicator stopAnimating];
}

- (void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [_activityIndicator stopAnimating];
}

- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [_activityIndicator stopAnimating];
    switch (error.code) {
        case NSURLErrorTimedOut:
            printf("timeout");
            break;
        case NSURLErrorNotConnectedToInternet:
            printf("no Internet");
            break;
        default:
            break;
    }
}
@end

File: DocScannerX/DocScannerX/UploadedFileViewController.swift
//
//  UploadedFileViewController.swift
//  Doc-Scanner-X
//
//  Created by dynamsoft on 2017/12/1.
//  Copyright © 2017年 Dynamsoft. All rights reserved.
//

import UIKit
import WebKit

class UploadedFileViewController: UIViewController, WKNavigationDelegate {

    // MARK: - Properties
    var webView: WKWebView!
    var navigationBar: UINavigationBar!
    let progressIndicator = UIActivityIndicatorView(activityIndicatorStyle: .gray)
    var refreshControl: UIRefreshControl?
    var navigationBarHeight: CGFloat {
        return (UIScreen.main.bounds.height == 812 ? 88 : 64)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        /// Initialize `webView`.
        let webConfiguration = WKWebViewConfiguration()
        webView = WKWebView(frame: .zero, configuration: webConfiguration)
        webView.allowsBackForwardNavigationGestures = true
        webView.navigationDelegate = self
        webView.frame = CGRect(x: 0, y: 64, width: fullScreenSize.width, height: fullScreenSize.height-navigationBarHeight)
        
        /// Initialize `navigationBar`.
        navigationBar = CustomNavigationBar()
        self.navigationItem.title = "Uploaded Files"
        navigationBar.pushItem(self.navigationItem, animated: true)
        let parentVC = parent as? MainViewController
        navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "icon_back")?.withRenderingMode(UIImageRenderingMode.alwaysOriginal), style: .plain, target: parentVC, action: #selector(parentVC?.backToHome))
        
        /// Initialize `refreshControl`.
        refreshControl = UIRefreshControl()
        refreshControl?.attributedTitle = NSAttributedString(string: "Pull to refresh")
        refreshControl?.addTarget(self, action: #selector(refreshWeb), for: .valueChanged)
        webView.scrollView.bounces = true
        webView.scrollView.addSubview(refreshControl!)
        
        view.frame = CGRect(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height)
        
        /// Add `navigationBar` and `webView` as subviews of `view`.
        view.addSubview(navigationBar)
        view.addSubview(webView)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    /// Refresh `webView` when user enter this view.
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        webView.addObserver(self, forKeyPath: "canGoBack", options: .new, context: nil)
        let myURL = URL(string: "https://your.upload.server"+KeyChainManager.readUUID()!)
        let myRequest = URLRequest(url: myURL!)
        webView.load(myRequest)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        webView.removeObserver(self, forKeyPath: "canGoBack")
    }
    
    func refreshWeb() {
        let myURL = URL(string: "https://xxxxxxx"+KeyChainManager.readUUID()!)
        let myRequest = URLRequest(url: myURL!)
        webView.load(myRequest)
        refreshControl?.endRefreshing()
    }
    
    func back() {
        webView.goBack()
    }
    
    // MARK: - Observer function
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if webView.canGoBack {
            navigationItem.leftBarButtonItem?.target = self
            navigationItem.leftBarButtonItem?.action = #selector(back)
        } else {
            let parentVC = parent as? MainViewController
            navigationItem.leftBarButtonItem?.target = parentVC
            navigationItem.leftBarButtonItem?.action = #selector(parentVC?.backToHome)
        }
    }
    
    // MARK: - WKNavigationDelegate
    
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        
        progressIndicator.backgroundColor = UIColor.clear
        progressIndicator.center = CGPoint(x: fullScreenSize.width/2, y: fullScreenSize.height/2)
        progressIndicator.hidesWhenStopped = true
        view.addSubview(progressIndicator)
        progressIndicator.startAnimating()
    }
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        progressIndicator.stopAnimating()
    }
    
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        progressIndicator.stopAnimating()
    }
    
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        progressIndicator.stopAnimating()
        if let err = error as? URLError {
            switch err.code {
            case URLError.timedOut:
                print("timeout")
            case URLError.notConnectedToInternet:
                print("no Internet")
            default:
                print("other error")
            }
        }
    }
}
